src/containerbuddy/etcd.go

```diff
@@ -0,0 +1,240 @@
+	return fmt.Sprintf("%s/%s/%s", c.Prefix, service.Name, service.ID)
+}
+
+func (c Etcd) getAppKey(appName string) string {
+	return fmt.Sprintf("%s/%s", c.Prefix, appName)
+}
+
+var etcdUpstreams = make(map[string][]EtcdServiceNode)
+
+func (c Etcd) checkHealth(backend *BackendConfig) bool {
+	services, err := c.getServices(backend.Name)
+	if err != nil {
+		if _, ok := err.(client.Error); !ok {
+			log.Printf("Failed to query %v: %s", backend.Name, err)
+		}
+		return false
+	}
+	didChange := etcdCompareForChange(etcdUpstreams[backend.Name], services)
+	if didChange || len(services) == 0 {
+		// We don't want to cause an onChange event the first time we read-in
+		// but we do want to make sure we've written the key for this map
+		etcdUpstreams[backend.Name] = services
+	}
+	return didChange
+}
+
+func (c Etcd) checkServiceExists(service *ServiceConfig) bool {
+	key := c.getNodeKey(service)
+	_, err := c.API.Get(context.Background(), key, nil)
+	if err != nil {
+		if etcdErr, ok := err.(client.Error); ok {
+			return etcdErr.Code != client.ErrorCodeKeyNotFound
+		}
+		log.Printf("Unexpected etcd Error on key %s: %s", key, err)
+	}
+	return true
+}
+
+func (c Etcd) getServices(appName string) ([]EtcdServiceNode, error) {
+	var services []EtcdServiceNode
+
+	key := c.getAppKey(appName)
+	resp, err := c.API.Get(context.Background(), key, &client.GetOptions{Recursive: true})
+	if err != nil {
+		if etcdErr, ok := err.(client.Error); ok {
+			if etcdErr.Code == client.ErrorCodeKeyNotFound {
+				return services, nil
+			}
+		}
+		log.Printf("Unable to get services: %s: %s", key, err)
+		return services, err
+	}
+	if !resp.Node.Dir {
+		log.Printf("Etcd key %s is not a directory", key)
+		return services, err
+	}
+	for _, instance := range resp.Node.Nodes {
+		if !instance.Dir {
+			continue
+		}
+		for _, node := range instance.Nodes {
+			if service, err := decodeEtcdNodeValue(node); err != nil {
+				log.Printf("Could not decode etcd service %s: %s", node.Value, err)
+			} else {
+				services = append(services, service)
+			}
+		}
+	}
+	return services, nil
+}
+
+// Compare the two arrays to see if the address or port has changed
+// or if we've added or removed entries.
+func etcdCompareForChange(existing, new []EtcdServiceNode) (changed bool) {
+	if len(existing) != len(new) {
+		return true
+	}
+
+	sort.Sort(ByEtcdServiceID(existing))
+	sort.Sort(ByEtcdServiceID(new))
+	for i, ex := range existing {
+		if ex.Address != new[i].Address ||
+			ex.Port != new[i].Port {
+			return true
+		}
+	}
+	return false
+}
+
+func (c Etcd) registerService(service *ServiceConfig) error {
+	key := c.getNodeKey(service)
+	value := encodeEtcdNodeValue(service)
+	ttl, _ := time.ParseDuration(fmt.Sprintf("%ds", service.TTL))
+	_, err := c.API.Set(context.Background(), key, "", &client.SetOptions{Dir: true, TTL: ttl})
+	if err != nil {
+		return err
+	}
+	_, err = c.API.Set(context.Background(), fmt.Sprintf("%s/%s", key, "/service"), value, nil)
```

Comment:
Is having to make two calls here potentially exposing us to getting into an inconsistent state if the second one fails? Will it just register safely on the next health check? We should probably have some notes here about what the expected behavior is.